cmd/dep/base_importer.go

```diff
@@ -0,0 +1,277 @@
+//
+// Rules:
+// * When a constraint is ignored, default to *.
+// * HEAD revisions default to the matching branch.
+// * Semantic versions default to ^VERSION.
+// * Revision constraints are ignored.
+// * Versions that don't satisfy the constraint, drop the constraint.
+// * Untagged revisions ignore non-branch constraint hints.
+func (i *baseImporter) importPackages(packages []importedPackage, defaultConstraintFromLock bool) (err error) {
+	projects, err := i.loadPackages(packages)
+	if err != nil {
+		return err
+	}
+
+	for _, prj := range projects {
+		pc := gps.ProjectConstraint{
+			Ident: gps.ProjectIdentifier{
+				ProjectRoot: prj.Root,
+				Source:      prj.Source,
+			},
+		}
+
+		pc.Constraint, err = i.sm.InferConstraint(prj.ConstraintHint, pc.Ident)
+		if err != nil {
+			pc.Constraint = gps.Any()
+		}
+
+		var version gps.Version
+		if prj.LockHint != "" {
+			var isTag bool
+			// Determine if the lock hint is a revision or tag
+			isTag, version, err = i.isTag(pc.Ident, prj.LockHint)
+			if err != nil {
+				return err
+			}
+
+			// If the hint is a revision, check if it is tagged
+			if !isTag {
+				revision := gps.Revision(prj.LockHint)
+				version, err = i.lookupVersionForLockedProject(pc.Ident, pc.Constraint, revision)
+				if err != nil {
+					version = nil
+					i.logger.Println(err)
+				}
+			}
+
+			// Default the constraint based on the locked version
+			if defaultConstraintFromLock && prj.ConstraintHint == "" && version != nil {
+				props := getProjectPropertiesFromVersion(version)
+				if props.Constraint != nil {
+					pc.Constraint = props.Constraint
+				}
+			}
+		}
+
+		// Ignore pinned constraints
+		if i.isConstraintPinned(pc.Constraint) {
+			if i.verbose {
+				i.logger.Printf("  Ignoring pinned constraint %v for %v.\n", pc.Constraint, pc.Ident)
+			}
+			pc.Constraint = gps.Any()
+		}
+
+		// Ignore constraints which conflict with the locked revision, so that
+		// solve doesn't later change the revision to satisfy the constraint.
+		if !i.testConstraint(pc.Constraint, version) {
+			if i.verbose {
+				i.logger.Printf("  Ignoring constraint %v for %v because it would invalidate the locked version %v.\n", pc.Constraint, pc.Ident, version)
+			}
+			pc.Constraint = gps.Any()
+		}
+
+		i.manifest.Constraints[pc.Ident.ProjectRoot] = gps.ProjectProperties{
+			Source:     pc.Ident.Source,
+			Constraint: pc.Constraint,
+		}
+		fb.NewConstraintFeedback(pc, fb.DepTypeImported).LogFeedback(i.logger)
+
+		if version != nil {
+			lp := gps.NewLockedProject(pc.Ident, version, nil)
+			i.lock.P = append(i.lock.P, lp)
+			fb.NewLockedProjectFeedback(lp, fb.DepTypeImported).LogFeedback(i.logger)
+		}
+	}
+
+	return nil
+}
+
+func (i *baseImporter) isConstraintPinned(c gps.Constraint) bool {
+	if version, isVersion := c.(gps.Version); isVersion {
+		switch version.Type() {
+		case gps.IsRevision:
+			return true
+		case gps.IsVersion:
+			return true
```

Comment:
nit:
```go
case gps.IsRevision, gps.IsVersion:
 return true
```

Reply:
Thanks! I knew there was a better way of doing that! ❤️